sdk/context.ts

```diff
@@ -0,0 +1,97 @@
+import { NetworkId } from '@synthetixio/contracts-interface';
+import { Provider as EthCallProvider } from 'ethcall';
+import { ethers } from 'ethers';
+
+import * as sdkErrors from './common/errors';
+import { ContractName, getContractsByNetwork, ContractMap } from './contracts';
+
+export interface IContext {
+	provider?: ethers.providers.Provider;
+	signer?: ethers.Signer;
+	networkId?: NetworkId;
+	walletAddress?: string;
+}
+
+const DEFAULT_CONTEXT: IContext = {
+	networkId: 10,
+};
+
+export default class Context implements IContext {
+	private context: IContext;
+	public multicallProvider = new EthCallProvider();
+	public contracts: ContractMap;
+
+	constructor(context: IContext) {
+		this.context = { ...DEFAULT_CONTEXT, ...context };
+
+		if (context.provider) {
+			this.multicallProvider.init(context.provider);
+		}
+
+		if (context.signer) {
+			this.setSigner(context.signer);
+		}
+
+		this.contracts = this.getContracts();
+	}
+
+	get networkId() {
+		return this.context.networkId ?? 10;
+	}
+
+	get provider() {
+		if (!this.context.provider) {
+			throw new Error(sdkErrors.NO_PROVIDER);
+		}
+
+		return this.context.provider;
+	}
+
+	get signer() {
+		if (!this.context.signer) {
+			throw new Error(sdkErrors.NO_SIGNER);
```

Comment:
Not sure about the pattern of throwing errors inside getters. Makes sense why it's been added so we don't have to keep throwing the same error inside functions but I think this can be a bit misleading with expecting a safe reference here.

Reply:
It's a little ugly, but it saves us from having to check for the existence of the values every time we want to use them. It also ensures type safety when we access the values.